### PR TITLE
Live: optionally send queries over websocket connection

### PR DESCRIPTION
**Pull request by FZambia**
**What this PR does / why we need it**:

One possible approach to solve streaming datasource HA issues when initial request goes to one node and subscribe goes to another is to query data over WS connection too.

This pr adds such functionality protected by a feature toggle `queryOverLive`. To reuse existing query logic this pr adds new `query.Service` – so HTTP API and Live can use it.

### CODE REVIEW

**Comment by ryantxu** [comment]
@ArturWierzbicki @FZambia -- I think this is close (for a POC!), but I am stuck with this compile error:
image

File: public/app/features/live/centrifuge/service.ts [hunk 1]
```diff
@@ -268,6 +275,35 @@ export class CentrifugeService implements CentrifugeSrv {
...
 
+  /**
+   * Execute a query over the live websocket and potentiall subscribe to a live channel.
+   *
+   * Since the initial request and subscription are on the same socket, this will support HA setups
+   */
+  getQueryData(options: LiveQueryDataOptions): Observable<DataQueryResponse> {
+    return new Observable((subscriber) => {
+      this.centrifuge
+        .rpc(options.body)
```
Comment (ryantxu) [inline public/app/features/live/centrifuge/service.ts:286]: @FZambia is this scoped at all?  Currently just passing the identical body here -- but I suspect we may want to support multiple methods

Comment (FZambia) [reply public/app/features/live/centrifuge/service.ts:286]: Yeah, there is a `namedRPC` method, which should become just an `rpc` in the next major release.

**Comment by ArturWierzbicki** [comment]
@ryantxu https://github.com/grafana/grafana/commit/ca3294692004d3807ba7e0722736b0153aa85b59 this should do the trick - fixed name & proxied remote observable

**Comment by ryantxu** [comment]
:facepalm: thanks!

**Comment by FZambia** [comment]
Just checked the following setup:
* 2 grafana nodes
* 1 load balancer
* signal datasource queries

Previously we could get 404 errors on dashboards since requests came to different nodes, with `queryOverLive` everything works expected.

**Comment by ryantxu** [comment]
@ArturWierzbicki with the feature flags:
```
[feature_toggles]
enable = live-pipeline, live-service-web-worker, queryOverLive
```

I get:
Peek 2021-11-30 08-51

image

**Comment by ArturWierzbicki** [comment]
hm, it does not work at all for me, because importing `toDataQueryResponse` from `@grafana/runtime`pulls in web app code to the web worker https://github.com/grafana/grafana/blob/9f10abd99a58fbb866da2b8e14ed2f356dae1495/public/app/features/live/centrifuge/service.ts#L2 

image
will take a look into it tom morning

**Comment by ryantxu** [comment]
In this case the webworker could just blindly pass the result payload back to the non-worker thread?  then we would not need `toDataQueryResponse` in the worker

**Comment by ArturWierzbicki** [comment]
> In this case the webworker could just blindly pass the result payload back to the non-worker thread? then we would not need `toDataQueryResponse` in the worker

yeah, we can do that but the subscription happens based on the properties of parsed object: https://github.com/grafana/grafana/blob/9f10abd99a58fbb866da2b8e14ed2f356dae1495/public/app/features/live/centrifuge/service.ts#L301

I guess we could return raw object (with a promise instead of an observable?), parse it in `GrafanaLiveService` and return observable if there are any subscribable channels?

**Comment by ArturWierzbicki** [comment]
implementing it rn

File: public/app/features/live/centrifuge/serviceWorkerProxy.ts [hunk 2]
```diff
@@ -28,6 +28,14 @@ export class CentrifugeServiceWorkerProxy implements CentrifugeSrv {
     );
   };
 
+  /**
+   * Query over websocket
+   */
+  getQueryData: CentrifugeSrv['getQueryData'] = async (options) => {
+    const optionsAsPlainSerializableObject = JSON.parse(JSON.stringify(options));
```
Comment (ArturWierzbicki) [inline public/app/features/live/centrifuge/serviceWorkerProxy.ts:35]: https://github.com/grafana/grafana/pull/41653#issuecomment-982825811 that was the issue. JSON parse+stringify is one option, the other one is just explicitly pick the properties we need and make sure they are plain&serializable @ryantxu

File: packages/grafana-runtime/src/services/live.ts [hunk 3]
```diff
@@ -20,12 +21,20 @@ export interface LiveDataFilter {
  */
 export interface LiveDataStreamOptions {
   addr: LiveChannelAddress;
-  frame?: DataFrame; // initial results
+  frame?: DataFrameJSON; // initial results
```
Comment (ArturWierzbicki) [inline packages/grafana-runtime/src/services/live.ts:24]: `options.frame` is passed from the main thread to web worker and `DataFrame` class is not serializable due to Vectors

File: packages/grafana-runtime/src/services/live.ts [hunk 4]
```diff
@@ -45,6 +54,15 @@ export interface GrafanaLiveSrv {
    */
   getDataStream(options: LiveDataStreamOptions): Observable<DataQueryResponse>;
 
+  /**
+   * Execute a query over the live websocket and potentiall subscribe to a live channel.
+   *
+   * Since the initial request and subscription are on the same socket, this will support HA setups
+   *
+   * NOTE: this function willl be called when the feature toggle `queryOverLive` is set
```
Comment (ryantxu) [inline packages/grafana-runtime/src/services/live.ts:62]:
```suggestion
   * @aplha -- this function requires the feature toggle `queryOverLive` is be set
```

File: public/app/features/live/centrifuge/serviceWorkerProxy.ts (hunk 2 above)
Comment (ryantxu) [reply public/app/features/live/centrifuge/serviceWorkerProxy.ts:35]: Is the main offender is timeRange?  

I think making the object more explicit is a better plan -- and exclude moment!

**Comment by mckn** [comment]
What do you think about the idea of keeping this out of the "application layer" and handle it in the infrastructure layer? E.g. making sure the user ends up on the same node for the current session.

**Comment by FZambia** [comment]
> What do you think about the idea of keeping this out of the "application layer" and handle it in the infrastructure layer? E.g. making sure the user ends up on the same node for the current session.

@mckn hi!

We are currently adding this behind feature toggle and we still need time to evaluate whether we benefit from going this way. So having stickness like this seems much more simple to evaluate. If for some use case this will make sense we may try to make it configurable on a more granular level rather than globally over feature toggle.

We have not really considered using sticky sessions yet – this way seems harder to evaluate and justify.

But we also looking for some alternatives at the moment where sending RPC over WS won't be required for streaming queries to properly work in HA setup. Far from ready, but – https://github.com/grafana/grafana/pull/43145. At the moment we have a working POC and some ideas on how to handle various failure scenarios. If you are interested I can tell more about all of this - if implemented it can make calling data over WS not needed at all.

**Review by ryantxu — APPROVED** [review]
I think this is ready to merge.  It works well and we can evaluate if it is an appropriate general option or something that can be done selectively.

Something we may need to revisit is the query inspector :face_with_head_bandage: it is just showing the last results that went though `getBackendSrv()`!

image

**Comment by ryantxu** [comment]
The breaking change detection is :fire: but in this case intentional and on an `@alpha` interface:

image

**Comment by ryantxu** [comment]
@leventebalogh -- is there a way to tell the bot that this is not a breaking change?  Do I just remove the label after merge?

**Comment by leventebalogh** [comment]
Thanks for raising it @ryantxu! 🙏 
(The workflow is still in an early state and rough around the edges)

**TL;DR**
**Just like you said, I think it's fine to remove the label and merge it.**

**Regarding this change**
I agree, as the `@alpha` channel is in the TSDoc I think it should not be marked as an actual breaking change.
I am a bit concerned though how much developers are aware of these TSDoc release tags and how much support their IDE gives them when they are trying to use APIs like this. Probably something for us (@grafana/plugins-platform-frontend) to investigate, as we would like to make (plugin) development as smooth as possible.

**How to approach these changes?**
We don't have any available guidelines _yet_, but we are working on it to make resolving these as easy as possible.
I think having a short discussion in the pull-request is generally a good idea, as it will help understanding the **why** behind certain changes for anyone who was not closely involved.
